### DISCUSSION TITLE
Recursive function calls

### DISCUSSION THREAD

**Discussion by adriangb** [geldata/gel#2101] [Ideas]
On a somewhat related note to #1764, would it be possible to support recursive function calls? Example use case:

```
type SomeThing {
    link child -> SomeThing;
 }

function get_all_children_ordered(parent: SomeThing) -> SET OF SomeThing {
        USING (
            SELECT child UNION get_all_children_ordered(parent)
        );
    }
```

This currently fails with

```
ERROR: InvalidReferenceError: function 'get_children_heirarchy' does not exist
```

But maybe it makes more sense to have recursive link traversal + aggregation functions + stopping conditions?

**Comment by elprans** [comment]
It should be possible to add support for recursive functions.  Creating one directly is fiddly, but doing `CREATE FUNCTION` with a non-recursive body first and the `ALTER`-ing it with a recursive body should work.  Unfortunately, it seems like this throws the DDL handler into an infinite loop.  I opened #2104 to track.


> But maybe it makes more sense to have recursive link traversal + aggregation functions + stopping conditions?

Yes, this is a popular request and there is a plan to implement support for recursive traversals.

**Reply by adriangb** [reply]
Thank you for taking a look.

I also experienced #2104 as well when trying to use `ALTER`.

> Yes, this is a popular request and there is a plan to implement support for recursive traversals.

Just for my learning: do you think that recursive transversals would replace the approach I propose above, or would it not cover that type of use case (in other words, what I proposed above is the right way to do this)?

**Reply by elprans** [reply]
I imagine that future recursive traversal syntax would be sufficient to implement queries like the above.
